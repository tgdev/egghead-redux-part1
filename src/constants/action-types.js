export const ACTION_ADD_TODO = 'ADD_TODO';
