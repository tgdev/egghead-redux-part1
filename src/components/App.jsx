import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { ACTION_ADD_TODO } from 'constants/action-types';
import store from '../store';

let nextTodoId = 0;

class TodoApp extends Component {

  render() {
    return (
      <div>
        <button
          onClick={() => {
            store.dispatch({
              type: ACTION_ADD_TODO,
              id: nextTodoId++,
              text: 'Test'
            });
          }}
        >
          Add Todo
        </button>
        <ul>
          {this.props.todos.map(todo => {
            return (
              <li key={todo.id}>
                {todo.text}
              </li>
            );
          })}
        </ul>
      </div>
    );
  }
}

TodoApp.propTypes = {
  todos: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    text: PropTypes.string.isRequired,
    completed: PropTypes.oneOf([true, false])
  })).isRequired
};

export default TodoApp;
