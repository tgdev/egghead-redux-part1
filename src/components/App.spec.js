import React from 'react';
import { shallow, mount } from 'enzyme';

import { FILTER_OPTION_ALL } from 'constants/visibility-filter-options';

import newTodosMock from 'mocks/new-todos.mock';
import updatedTodosMock from 'mocks/updated-todos.mock';

import App from './App';

describe('App component', () => {

  const mockTodos = newTodosMock;
  const altMockTodos = updatedTodosMock;

  const mockVisibiltyFilter = FILTER_OPTION_ALL;

  it('renders TodoApp correctly with no todos', () => {
    const wrapper = shallow(<App todos={[]} visibilityFilter={mockVisibiltyFilter} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.instance().props.todos.length).toEqual(0);
  });

  it('renders TodoApp correctly with saved todos', () => {
    const wrapper = shallow(<App todos={mockTodos} visibilityFilter={mockVisibiltyFilter} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.instance().props.todos.length).toEqual(2);
  });

  it('renders a completed todo with line-through', () => {
    const wrapper = shallow(<App todos={altMockTodos} visibilityFilter={mockVisibiltyFilter} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('triggers handleAddTodo when "Add Todo" button is clicked', () => {
    const handleAddTodo = jest.spyOn(App.prototype, 'handleAddTodo');
    const wrapper = mount(<App todos={[]} visibilityFilter={mockVisibiltyFilter} />);
    wrapper.find('button').first().simulate('click');
    expect(handleAddTodo).toHaveBeenCalledTimes(1);
  });

  it('triggers handleToggleTodo when todo item is clicked', () => {
    const handleToggleTodo = jest.spyOn(App.prototype, 'handleToggleTodo');
    const wrapper = mount(<App todos={mockTodos} visibilityFilter={mockVisibiltyFilter} />);
    wrapper.find('li').first().simulate('click');
    expect(handleToggleTodo).toHaveBeenCalledTimes(1);
  });

});
