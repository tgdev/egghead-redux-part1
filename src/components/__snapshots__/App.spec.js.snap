// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders TodoApp correctly with no todos 1`] = `
<div>
  <Connect(AddTodo) />
  <Connect(TodoList) />
  <TodoFooter />
</div>
`;

exports[`App component renders TodoApp correctly with saved todos 1`] = `
<div>
  <Connect(AddTodo) />
  <Connect(TodoList) />
  <TodoFooter />
</div>
`;
